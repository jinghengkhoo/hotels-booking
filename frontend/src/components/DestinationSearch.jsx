import { useState, useEffect } from "react";
import axios from "axios";
import { useNavigate } from "react-router-dom";
import "react-datepicker/dist/react-datepicker.css";
import { useCombobox } from "downshift";
import CustomDatePicker from "./DatePicker";

const DestinationSearch = () => {
  const [destination, setDestination] = useState("");
  const [destinationId, setDestinationId] = useState("");
  const [suggestions, setSuggestions] = useState([]);
  const [startDate, setStartDate] = useState(null);
  const [endDate, setEndDate] = useState(null);
  const [guests, setGuests] = useState(1);
  const [rooms, setRooms] = useState(1);
  const navigate = useNavigate();

  useEffect(() => {
    if (destination.length > 0) {
      axios
        .get(`http://localhost:5555/api/destinations?query=${destination}`)
        .then((res) => {
          setSuggestions(res.data);
        })
        .catch((err) => console.error(err));
    } else {
      setSuggestions([]);
    }
  }, [destination]);

  const validateForm = () => {
    const newErrors = {};

    if (!destinationId) {
      newErrors.destinationId = "Please select a valid destination.";
    }
    if (!startDate) {
      newErrors.startDate = "Please select a start date.";
    }
    if (!endDate) {
      newErrors.endDate = "Please select an end date.";
    }
    if (guests <= 0) {
      newErrors.guests = "Please select the number of guests.";
    }
    if (rooms <= 0) {
      newErrors.rooms = "Please select the number of rooms.";
    }

    return Object.keys(newErrors).length === 0;
  };

  const handleFormSubmit = (e) => {
    e.preventDefault();
    if (validateForm()) {
      navigate("/hotels", {
        state: { destinationId, startDate, endDate, guests, rooms },
      });
    }
  };

  const {
    isOpen,
    getMenuProps,
    getInputProps,
    highlightedIndex,
    getItemProps,
  } = useCombobox({
    onInputValueChange: ({ inputValue }) => {
      setDestination(inputValue);
    },
    onSelectedItemChange: ({ selectedItem }) => {
      setDestination(selectedItem ? selectedItem.term : "");
      setDestinationId(selectedItem ? selectedItem.uid : "");
    },
    items: suggestions,
    itemToString: (item) => (item ? item.term : ""),
  });

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-100">
      <form
        className="bg-white p-8 rounded shadow-md w-full max-w-md"
        onSubmit={handleFormSubmit}
      >
        <h2 className="text-2xl mb-6">Search Destination</h2>
        <div className="mb-4">
          <label className="block text-gray-700">Destination</label>
          <div className="relative">
            <div className="flex shadow-sm bg-white gap-0.5">
              <input
                placeholder="Type a destination"
                className="mt-1 block w-full px-3 py-2 bg-white border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
                {...getInputProps()}
                type="text"
                name="destination"
              />
            </div>

            <ul
              className={`absolute w-72 bg-white mt-1 shadow-md max-h-80 overflow-scroll p-0 z-10 ${
                !(isOpen && suggestions.length) && "hidden"
              }`}
              {...getMenuProps()}
            >
              {isOpen &&
                suggestions.map((item, index) => (
                  <li
                    key={item.id}
                    {...getItemProps({ item, index })}
                    className={`cursor-pointer px-4 py-2 ${
                      highlightedIndex === index
                        ? "bg-blue-500 text-white"
                        : "bg-white text-black"
                    }`}
                  >
                    {item.term}
                  </li>
                ))}
            </ul>
          </div>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700">Start Date</label>
          <CustomDatePicker selectedDate={startDate} onChange={setStartDate} />
        </div>
        <div className="mb-4">
          <label className="block text-gray-700">End Date</label>
          <CustomDatePicker selectedDate={endDate} onChange={setEndDate} />
        </div>
        <div className="mb-4">
          <label className="block text-gray-700">Guests</label>
          <select
            value={guests}
            onChange={(e) => setGuests(Number(e.target.value))}
            className="mt-1 block w-full px-3 py-2 bg-white border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
          >
            {[...Array(10).keys()].map((i) => (
              <option key={i + 1} value={i + 1}>
                {i + 1} {i + 1 === 1 ? "Guest" : "Guests"}
              </option>
            ))}
          </select>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700">Rooms</label>
          <select
            value={rooms}
            onChange={(e) => setRooms(Number(e.target.value))}
            className="mt-1 block w-full px-3 py-2 bg-white border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
          >
            {[...Array(10).keys()].map((i) => (
              <option key={i + 1} value={i + 1}>
                {i + 1} {i + 1 === 1 ? "Room" : "Rooms"}
              </option>
            ))}
          </select>
        </div>
        <button
          type="submit"
          className="w-full py-2 px-4 bg-blue-600 hover:bg-blue-700 text-white font-semibold rounded-lg shadow-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-opacity-75"
        >
          Search
        </button>
      </form>
    </div>
  );
};

export default DestinationSearch;
