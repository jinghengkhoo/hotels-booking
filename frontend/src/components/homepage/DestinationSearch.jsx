import "react-datepicker/dist/react-datepicker.css";
import singaporeImage from "../../assets/singapore.jpg"; // Correct import for the image
import SearchBar from "../SearchBar"; // Ensure the path is correct

const DestinationSearch = () => {
  return (
    <div
      className="min-h-screen flex flex-col items-center justify-center bg-cover bg-center h-3/4"
      style={{ backgroundImage: `url(${singaporeImage})` }}
    >
      <div className="absolute inset-0 bg-black opacity-20 mt-20"></div>
      <div className="mt-64">
        <div className="relative z-10 text-white text-center p-8">
          <h1 className="text-4xl font-bold">
            Visit the best places in the world with us.
          </h1>
          <p className="text-xl mt-2">Book your hotel with us now.</p>
        </div>
        <div className="relative z-10">
          <SearchBar />
        </div>
      </div>
    </div>
  );
};

export default DestinationSearch;
