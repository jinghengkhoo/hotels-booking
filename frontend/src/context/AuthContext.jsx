import { createContext, useState, useEffect } from "react";
import axios from "axios";
import PropTypes from "prop-types";

const AuthContext = createContext();

const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  useEffect(() => {
    const checkLoginStatus = async () => {
      try {
        const res = await axios.get("http://localhost:5555/api/user/", {
          withCredentials: true,
        });
        setUser(res.data);
      } catch (err) {
        console.log(err);
        setUser(null);
      }
    };

    checkLoginStatus();
  }, []);

  const logout = async () => {
    await axios.post(
      "http://localhost:5555/api/user/logout",
      {},
      { withCredentials: true }
    );
    setUser(null);
  };

  return (
    <AuthContext.Provider value={{ user, setUser, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export { AuthContext, AuthProvider };
