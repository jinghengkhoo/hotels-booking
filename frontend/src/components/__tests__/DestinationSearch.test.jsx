import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import DestinationSearch from '../homepage/DestinationSearch';
import singaporeImage from '../../assets/singapore.jpg'; // Correct import for the image
import { AuthContext } from '../../context/AuthContext';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

jest.mock('../NavBar', () => () => <div data-testid="navbar">NavBar Mock</div>);
jest.mock('../SearchBar', () => () => <div data-testid="searchbar">SearchBar Mock</div>);

// Mock useNavigate hook
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
}));

const mockAuthContextValue = {
  user: { name: 'Test User' },
  logout: jest.fn(),
};

describe('DestinationSearch', () => {
  test('renders NavBar, background image, heading, and SearchBar', () => {
    render(
      <AuthContext.Provider value={mockAuthContextValue}>
        <Router>
          <DestinationSearch />
        </Router>
      </AuthContext.Provider>
    );

    // Check if NavBar is rendered
    const navBarElement = screen.getByTestId('navbar');
    expect(navBarElement).toBeInTheDocument();

    // Check if the background image is set correctly
    const mainElement = screen.getByRole('main');
    expect(mainElement).toHaveStyle(`background-image: url(${singaporeImage})`);

    // Check if heading is rendered
    const headingElement = screen.getByText('Singapore, Singapore');
    expect(headingElement).toBeInTheDocument();

    // Check if the subheading is rendered
    const subheadingElement = screen.getByText('Get ready for your next adventure â†’');
    expect(subheadingElement).toBeInTheDocument();

    // Check if SearchBar is rendered
    const searchBarElement = screen.getByTestId('searchbar');
    expect(searchBarElement).toBeInTheDocument();
  });
  test('matches snapshot', () => {
    const tree = renderer.create(
      <AuthContext.Provider value={mockAuthContextValue}>
        <Router>
          <DestinationSearch />
        </Router>
      </AuthContext.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
