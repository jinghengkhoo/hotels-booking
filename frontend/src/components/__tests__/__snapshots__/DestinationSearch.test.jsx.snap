// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationSearch matches snapshot 1`] = `
<div
  className="min-h-screen flex flex-col bg-base-100"
>
  <main
    className="w-full flex flex-col items-center bg-cover bg-center relative main-background opacity-100"
    style={
      {
        "backgroundImage": "url([object Object])",
        "height": "80vh",
      }
    }
  >
    <div
      data-testid="navbar"
    >
      NavBar Mock
    </div>
    <div
      className="absolute inset-0 bg-black bg-opacity-20 py-10 px-48"
    >
      <div
        className="absolute bottom-0 left-0 mb-10 ml-48 text-transition opacity-100"
      >
        <h2
          className="text-4xl font-bold text-white mb-4"
        >
          Singapore, Singapore
        </h2>
        <p
          className="text-lg text-white mb-6"
        >
          Get ready for your next adventure â†’
        </p>
      </div>
    </div>
  </main>
  <div
    className="relative z-10 w-full max-w-screen-lg mx-auto mt-[-4rem] px-4"
  >
    <div
      data-testid="searchbar"
    >
      SearchBar Mock
    </div>
  </div>
</div>
`;
